### ISSUE TITLE
[HOLD for payment 2023-11-29] [$500] Distance Request – The map is zoomed incorrectly after adding description

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#29469]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.83.1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @paultsimura
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1697062856852229

## Action Performed:
1. Click FAB -> 'Request Money'
2. Choose 'Distance'
3. Add a route from point 1 to point 2:
- 123 W Main St, Independence, KS 67301, USA
- 321 Westminster Pl, Independence, KS 67301, USA
4. Click on the "Next" button
5. Select any WS
6. Verify that the whole route is within the map boundaries
7. Click "Description"
8. Move back

## Expected Result:
The route is within the map boundaries.

## Actual Result:
The map is zoomed a bit too much, the route doesn't fit the map.

## Workaround:
Unknown 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

 

https://github.com/Expensify/App/assets/93399543/ed9d8e7f-59f5-4e46-92d9-c6b23237975d

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

 

https://github.com/Expensify/App/assets/93399543/57ae72ec-7826-4d17-b1c1-3d0297309e2b

https://github.com/Expensify/App/assets/93399543/893eadbc-b1d7-43a9-86ef-328cbcbd90ec

</details>

<details>
<summary>MacOS: Desktop</summary>

 

https://github.com/Expensify/App/assets/93399543/943a6c61-dfb7-4593-8aea-fbd34f848c00

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~011040976b36d481db</li>
        <li>Upwork Job ID: 1712503476731002880</li>
        <li>Last Price Increase: 2023-10-12</li>
<li>Automatic offers: </li>
<ul>
<li>paultsimura | Contributor | 27595967</li>
<li>paultsimura | Reporter | 27595970</li>
</ul></li>
    </ul>
</details>

**Comment by kbecciv** [comment]
Proposal by: @paultsimura
Slack conversation: https://expensify.slack.com/archives/C049HHMV9SM/p1697062856852229


## Proposal
### Please re-state the problem that we are trying to solve in this issue.
After opening "Description" editor on Distance Request confirmation, the map is incorrectly zoomed.
### What is the root cause of that problem?
The root cause of this is that we use the ResizeObserver to resize the map. It's a good thing that was added in https://github.com/Expensify/App/pull/27574 to fix the map being cut off.
However, we are facing a race condition when opening a map back from the "Distance" tab.
First, the boundaries are reset by this hook:
https://github.com/Expensify/App/blob/05ab7ed1920b2e55f517cea7dc73c16f13d8fac7/src/components/MapView/MapView.web.tsx#L27-L51
Then, `mapRef.resize()` is called here:
https://github.com/Expensify/App/blob/05ab7ed1920b2e55f517cea7dc73c16f13d8fac7/src/components/MapView/MapView.web.tsx#L59
So the map is being resized after the boundaries have been reset.
### What changes do you think we should make in order to solve the problem?
First, we should split this hook into a function `resetBoundaries()`, and call it in the hook.
```js
        const resetBoundaries = useCallback(() => {
            ...current hook content here...
        }, [waypoints, mapRef, mapPadding, directionCoordinates]);
        useEffect(resetBoundaries, [resetBoundaries]);
```
https://github.com/Expensify/App/blob/05ab7ed1920b2e55f517cea7dc73c16f13d8fac7/src/components/MapView/MapView.web.tsx#L27-L51
And then, we should call `resetBoundaries()` after the resize:
```diff
            const resizeObserver = new ResizeObserver(() => {
                mapRef.resize();
+               resetBoundaries();
            });
```
https://github.com/Expensify/App/blob/05ab7ed1920b2e55f517cea7dc73c16f13d8fac7/src/components/MapView/MapView.web.tsx#L58-L60C16
#### Result:
https://user-images.githubusercontent.com/12595293/274435387-c286a5c8-9789-4fd9-98c7-025705352aef.mov
**P.s.** Since the resize happens only on a tab change, this will be a safe addition, which will also solve the issue of not resetting the zoom of the map when navigating back to the initial waypoints selection page. This one:
https://user-images.githubusercontent.com/12595293/274653852-d71fe309-cdd9-4508-9d28-60a13accb565.mov
### What alternative solutions did you explore? (Optional)

**Comment by paultsimura** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
After opening "Description" editor on Distance Request confirmation, the map is incorrectly zoomed.

### What is the root cause of that problem?
The root cause of this is that we use the ResizeObserver to resize the map. It's a good thing that was added in https://github.com/Expensify/App/pull/27574 to fix the map being cut off.

However, we are facing a race condition when opening a map back from the "Distance" tab.

First, the boundaries are reset by this hook:

https://github.com/Expensify/App/blob/05ab7ed1920b2e55f517cea7dc73c16f13d8fac7/src/components/MapView/MapView.web.tsx#L27-L51

Then, `mapRef.resize()` is called here:

https://github.com/Expensify/App/blob/d2c8456087be19607bd1b0075755936b8d438de8/src/components/MapView/MapView.web.tsx#L53-L66

So the map is being resized after the boundaries have been reset.

### What changes do you think we should make in order to solve the problem?
First, we should split this hook into a function `resetBoundaries()`, and call it in the hook.

```js
        const resetBoundaries = useCallback(() => {
            ...current hook content here...
        }, [waypoints, mapRef, mapPadding, directionCoordinates]);

        useEffect(resetBoundaries, [resetBoundaries]);
```

https://github.com/Expensify/App/blob/05ab7ed1920b2e55f517cea7dc73c16f13d8fac7/src/components/MapView/MapView.web.tsx#L27-L51

And then, we should call `resetBoundaries()` after the resize:

```diff
            const resizeObserver = new ResizeObserver(() => {
                mapRef.resize();
+               resetBoundaries();
            });
```


https://github.com/Expensify/App/blob/d2c8456087be19607bd1b0075755936b8d438de8/src/components/MapView/MapView.web.tsx#L58-L60


#### Result:

https://user-images.githubusercontent.com/12595293/274435387-c286a5c8-9789-4fd9-98c7-025705352aef.mov


**P.s.** Since the resize happens only on a tab change, this will be a safe addition, which will also solve the issue of not resetting the zoom of the map when navigating back to the initial waypoints selection page. This one:

https://user-images.githubusercontent.com/12595293/274653852-d71fe309-cdd9-4508-9d28-60a13accb565.mov



### What alternative solutions did you explore? (Optional)

P.p.s. thank you @kbecciv, I've reposted my proposal just to keep better formatting, as Slack messes it up a little bit.

**Comment by NicMendonca** [comment]
@Santhosh-Sellavel proposals for ya!

**Comment by Santhosh-Sellavel** [comment]
@kbecciv @NicMendonca Can you check again, this seems to be working fine already!

**Comment by paultsimura** [comment]
> @kbecciv @NicMendonca Can you check again, this seems to be working fine already!

@Santhosh-Sellavel did you test on Web/mWeb? Native platforms use different SDK and it's not reproducible there.

Also, this was reproduced just now:
https://github.com/Expensify/App/assets/12595293/96e1c3c8-612f-4b47-a7fa-7d1240361d98

**Comment by Santhosh-Sellavel** [comment]
mWeb

**Comment by paultsimura** [comment]
Please also make sure to use the points that are relatively close (the ones from the issue description are a perfect match). The zoom lag on large distances is not that noticeable.

@Santhosh-Sellavel

**Comment by Santhosh-Sellavel** [comment]
Alright got it, thanks @paultsimura!

**Comment by Santhosh-Sellavel** [comment]
@paultsimura [Proposal](https://github.com/Expensify/App/issues/29469#issuecomment-1760030800) looks good to me!

**Comment by Santhosh-Sellavel** [comment]
@paultsimura [Proposal](https://github.com/Expensify/App/issues/29469#issuecomment-1760030800) looks good to me!

C+ Reviewed
👀 🎀 👀

**Comment by Santhosh-Sellavel** [comment]
@paultsimura https://github.com/Expensify/App/issues/29469#issuecomment-1760030800 looks good to me!

C+ Reviewed
🎀 👀 🎀

**Comment by situchan** [comment]
I suggest to put this on hold for #26724 which was on hold for #28911

**Comment by deetergp** [comment]
I agree with @situchan, that this should go into the hold chain. https://github.com/Expensify/App/issues/28911 was merged earlier this morning and should go out to Staging once the current round of QA on staging is done.

**Comment by jliexpensify** [comment]
@deetergp - I've just closed #26724 based off @neil-marcellini 's [comments](https://github.com/Expensify/App/issues/26724#issuecomment-1775569926). I've also done some testing of the map zooming in that GH and we both think it's not a big deal.

**Comment by paultsimura** [comment]
I think this one is a bit different UX because it doesn't require any extra map interactions from the user: no previous zoom-in is required, neither any window screen resize.
It's just this: now you have a properly displayed route that fits the boundaries, and right after - you have a route that doesn't fit, even though you didn't touch the map.

cc: @neil-marcellini isn't this case worth fixing either?

**Comment by neil-marcellini** [comment]
Yes I agree with @paultsimura this is definitely a bug. It's super weird and his proposal to fix it looks great.

**Comment by deetergp** [comment]
Okay, let's go with @paultsimura's solution 👍

**Comment by paultsimura** [comment]
Thanks, the PR is ready: https://github.com/Expensify/App/pull/31156
@NicMendonca could you please remove the "Hold..." from the issue title? 🙏

**Comment by paultsimura** [comment]
@deetergp a friendly bump on the payout here, please☝️

**Comment by NicMendonca** [comment]
BZ Payment summary: 
Reporter: @paultsimura - $50 
Contributor: @paultsimura - $500 
C+ @Santhosh-Sellavel - $500 

@Santhosh-Sellavel don't forget to request payment via Expensify!

@paultsimura you've been paid via Upwork!

**Comment by NicMendonca** [comment]
@Santhosh-Sellavel bump on this! https://github.com/Expensify/App/issues/29469#issuecomment-1821913055

**Comment by NicMendonca** [comment]
@Santhosh-Sellavel [bump bump ](https://expensify.slack.com/archives/C02NK2DQWUX/p1702305130876669)

**Comment by Santhosh-Sellavel** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
> 
> * [x]  [@Santhosh-Sellavel] The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/27574
> * [x]  [@Santhosh-Sellavel] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/27574#issuecomment-1850624066
> * [x]  [@Santhosh-Sellavel] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: Should have been caught
> * [x]  [@Santhosh-Sellavel] Determine if we should create a regression test for this bug. Yes
> * [x]  [@Santhosh-Sellavel] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
> * [ ]  [@NicMendonca] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by NicMendonca** [comment]
Thanks @Santhosh-Sellavel!


> BZ Payment summary:
> Reporter: @paultsimura - $50
> Contributor: @paultsimura - $500
> C+ @Santhosh-Sellavel - $500
> 
> @Santhosh-Sellavel don't forget to request payment via Expensify!
> 
> @paultsimura you've been paid via Upwork!

**Comment by Santhosh-Sellavel** [comment]
Request On ND

**Comment by JmillsExpensify** [comment]
$500 payment approved for @Santhosh-Sellavel based on comment above.
